File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.py
class Solution:
    def findCheapestPrice(self, n, arr, src, dst, k):
        adj = defaultdict(list)
        for s, d, p in arr:
            adj[s].append((d, p))
        queue = deque()
        queue.append((src, 0))
        costs = [float("inf")] * n
        while queue and k >= 0:
            m = len(queue)
            for _ in range(m):
                curr, cost = queue.popleft()
                for node, price in adj[curr]:
                    if cost + price < costs[node]:
                        costs[node] = cost + price
                        queue.append((node, costs[node]))
            k -= 1
        return costs[dst] if costs[dst] != float("inf") else -1

File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& arr, int src, int dst, int k) {
        vector<vector<pair<int, int>>> adj(n);
        for(auto x : arr){
            adj[x[0]].push_back({x[1], x[2]});
        }
        queue<pair<int, int>> q;
        q.push({src, 0});
        vector<int> minCost(n, INT_MAX);
        int stops = 0;
        while(!q.empty() && stops <= k){
            int size = q.size();
            while (size--) {
                auto [curr, cost] = q.front();
                q.pop();
                for (auto& [neigh, price] : adj[curr]) {
                    if (price + cost < minCost[neigh]){
                        minCost[neigh] = price + cost;
                        q.push({neigh, minCost[neigh]});
                    }
                }
            }
            stops++;
        }
        if(minCost[dst] == INT_MAX)
            return -1;
        return minCost[dst];
    }
};